### ISSUE TITLE
Getting System Out Of Memory Exception while saving XLWorkBook to MemoryStream for datatable with 300K reacords.

### ISSUE THREAD

**Issue by shravandivity** [ClosedXML/ClosedXML#1753]
## Read and complete the full issue template

Do not randomly delete sections. They are here for a reason.

**Do you want to request a *feature* or report a *bug*?**
- [x] Bug
- [ ] Feature
- [ ] Question

**Did you test against the latest CI build?**
- [x] Yes
- [ ] No

If you answered `No`, please test with [the latest development build](https://ci.appveyor.com/project/ClosedXML/ClosedXML/branch/develop/artifacts) first.

**Version of ClosedXML**
0.95.4

**What is the current behavior?**
Getting out of memory exception while saving XlWorkBook to memory stream for datatable with 300K + records.

**What is the expected behavior or new feature?**

Complete this.

**Is this a regression from the previous version?**

Regressions get higher priority. Test against the latest build of the previous minor version. For example, if you experience a problem on v0.95.3, check whether it the problem occurred in v0.94.2 too. 

## Reproducibility
**This is an important section. Read it carefully. Failure to do so will cause a 'RTFM' comment.**

Without a code sample, it is unlikely that your issue will get attention. Don't be lazy. Do the effort and assist the developers to reproduce your problem. Code samples should be [minimal complete and verifiable](https://stackoverflow.com/help/mcve). Sample spreadsheets should be attached whenever applicable. Remove sensitive information.

**Code to reproduce problem:**
```c#
public void Main()
{
    DataTable dt = new DataTable();
DataColumn dc1 = new DataColumn("A", typeof(String));
DataColumn dc2 = new DataColumn("B", typeof(String));
DataColumn dc3 = new DataColumn("C", typeof(String));
DataColumn dc4 = new DataColumn("D", typeof(String));
DataColumn dc5 = new DataColumn("E", typeof(String));
DataColumn dc6 = new DataColumn("F", typeof(String));
DataColumn dc7 = new DataColumn("G", typeof(String));
DataColumn dc8 = new DataColumn("H", typeof(String));
DataColumn dc9 = new DataColumn("I", typeof(String));
DataColumn dc10 = new DataColumn("J", typeof(String));
DataColumn dc11 = new DataColumn("K", typeof(String));
DataColumn dc12 = new DataColumn("L", typeof(String));
DataColumn dc13 = new DataColumn("M", typeof(String));
DataColumn dc14 = new DataColumn("N", typeof(String));
DataColumn dc15 = new DataColumn("O", typeof(String));
DataColumn dc16 = new DataColumn("P", typeof(String));
DataColumn dc17 = new DataColumn("Q", typeof(String));
DataColumn dc18 = new DataColumn("R", typeof(String));
DataColumn dc19 = new DataColumn("S", typeof(String));
DataColumn dc20 = new DataColumn("T", typeof(String));
DataColumn dc21 = new DataColumn("U", typeof(String));
DataColumn dc22 = new DataColumn("V", typeof(String));
DataColumn dc23 = new DataColumn("W", typeof(String));
DataColumn dc24 = new DataColumn("X", typeof(String));
DataColumn dc25 = new DataColumn("Y", typeof(String));
DataColumn dc26 = new DataColumn("Z", typeof(String));
DataColumn dc27 = new DataColumn("AA", typeof(String));
DataColumn dc28 = new DataColumn("AB", typeof(String));
DataColumn dc29 = new DataColumn("AC", typeof(String));
DataColumn dc30 = new DataColumn("AD", typeof(String));
DataColumn dc31 = new DataColumn("AE", typeof(String));
DataColumn dc32 = new DataColumn("AF", typeof(String));
DataColumn dc33 = new DataColumn("AG", typeof(String));
DataColumn dc34 = new DataColumn("AH", typeof(String));
DataColumn dc35 = new DataColumn("AI", typeof(String));
DataColumn dc36 = new DataColumn("AJ", typeof(String));
DataColumn dc37 = new DataColumn("AK", typeof(String));
DataColumn dc38 = new DataColumn("AL", typeof(String));
DataColumn dc39 = new DataColumn("AM", typeof(String));
DataColumn dc40 = new DataColumn("AN", typeof(String));
DataColumn dc41 = new DataColumn("AO", typeof(String));
DataColumn dc42 = new DataColumn("AP", typeof(String));
DataColumn dc43 = new DataColumn("AQ", typeof(String));
DataColumn dc44 = new DataColumn("AR", typeof(String));
DataColumn dc45 = new DataColumn("AS", typeof(String));

dt.Columns.Add(dc1);
dt.Columns.Add(dc2);
dt.Columns.Add(dc3);
dt.Columns.Add(dc4);
dt.Columns.Add(dc5);
dt.Columns.Add(dc6);
dt.Columns.Add(dc7);
dt.Columns.Add(dc8);
dt.Columns.Add(dc9);
dt.Columns.Add(dc10);
dt.Columns.Add(dc11);
dt.Columns.Add(dc12);
dt.Columns.Add(dc13);
dt.Columns.Add(dc14);
dt.Columns.Add(dc15);
dt.Columns.Add(dc16);
dt.Columns.Add(dc17);
dt.Columns.Add(dc18);
dt.Columns.Add(dc19);
dt.Columns.Add(dc20);
dt.Columns.Add(dc21);
dt.Columns.Add(dc22);
dt.Columns.Add(dc23);
dt.Columns.Add(dc24);
dt.Columns.Add(dc25);
dt.Columns.Add(dc26);
dt.Columns.Add(dc27);
dt.Columns.Add(dc28);
dt.Columns.Add(dc29);
dt.Columns.Add(dc30);
dt.Columns.Add(dc31);
dt.Columns.Add(dc32);
dt.Columns.Add(dc33);
dt.Columns.Add(dc34);
dt.Columns.Add(dc35);
dt.Columns.Add(dc36);
dt.Columns.Add(dc37);
dt.Columns.Add(dc38);
dt.Columns.Add(dc39);
dt.Columns.Add(dc40);
dt.Columns.Add(dc41);
dt.Columns.Add(dc42);
dt.Columns.Add(dc43);
dt.Columns.Add(dc44);
dt.Columns.Add(dc45);

for(int i = 0; i < 300000; i++)
{
DataRow dr = dt.NewRow();
dr[0] = "Sample text to test";
dr[1] = "Sample text to test";
dr[2] = "Sample text to test";
dr[3] = "Sample text to test";
dr[4] = "Sample text to test";
dr[5] = "Sample text to test";
dr[6] = "Sample text to test";
dr[7] = "Sample text to test";
dr[8] = "Sample text to test";
dr[9] = "Sample text to test";
dr[10] = "Sample text to test";
dr[11] = "Sample text to test";
dr[12] = "Sample text to test";
dr[13] = "Sample text to test";
dr[14] = "Sample text to test";
dr[15] = "Sample text to test";
dr[16] = "Sample text to test";
dr[17] = "Sample text to test";
dr[18] = "Sample text to test";
dr[19] = "Sample text to test";
dr[20] = "Sample text to test";
dr[21] = "Sample text to test";
dr[22] = "Sample text to test";
dr[23] = "Sample text to test";
dr[24] = "Sample text to test";
dr[25] = "Sample text to test";
dr[26] = "Sample text to test";
dr[27] = "Sample text to test";
dr[28] = "Sample text to test";
dr[29] = "Sample text to test";
dr[30] = "Sample text to test";
dr[31] = "Sample text to test";
dr[32] = "Sample text to test";
dr[33] = "Sample text to test";
dr[34] = "Sample text to test";
dr[35] = "Sample text to test";
dr[36] = "Sample text to test";
dr[37] = "Sample text to test";
dr[38] = "Sample text to test";
dr[39] = "Sample text to test";
dr[40] = "Sample text to test";
dr[41] = "Sample text to test";
dr[42] = "Sample text to test";
dr[43] = "Sample text to test";
dr[44] = "Sample text to test";

dt.Rows.Add(dr);
}

XLWorkbook wb = new XLWorkbook();
var ws = wb.Worksheets.Add(dt, "Sheet1");
ws.Table(0).ShowAutoFilter = false;
ws.Table(0).Theme = XLTableTheme.None;
MemoryStream ms = new MemoryStream();
wb.SaveAs(ms);
}
```

- [ ] I attached a sample spreadsheet.  (You can drag files on to this issue)

**Comment by jahav** [comment]
You already filled the in the #1753. Reason is still the same:

* ClosedXml takes a memory, that takes up ~X memory
* When saving, it converts it to DocumentFormat.OpenXml, that takes about 1.5X 2X memory
* Saving to memory stream takes another several hundreds MBs memory.

Closing as duplicate #1753.

**Comment by igitur** [comment]
A duplicate of itself? 😜

**Comment by jahav** [comment]
Oops. Picked a wrong tab. Both issues have same author, problem code. .. good catch, thanks.

**Comment by swloo618** [comment]
How to over come this?

**Comment by jahav** [comment]
If you have problems with memory, profiling is your best friend to find what can be done.

As far as ClosedXml goes, I don't think there is way to shave memory at this time. I made an exploratory ticket #1874 , but don't hold your breath for resolution any time soon (if even possible).

Either try to buy more memory or try to shave other memory (e.g. instead of memory stream, save to to file/blob if you are on AWS/Azure and then stream response from there).

In the other report, you had pretty suspicious `Response.Buffer=true`
```
Response.Clear()
            Response.Buffer = True
            Response.Charset = ""
```

**Comment by jahav** [comment]
I have modified saving code, so it uses streaming of XML instead of saving the DOM in the #1874. I used the code from the original post (limited to 30k rows instead) and the results are at https://github.com/ClosedXML/ClosedXML/issues/1874#issuecomment-1355745624. 

Basically ClosedXML now only needs additional memory for the zip archive stream (the xlsx file). It doesn't build DOM that required so much additional memory (at least for cell values and string table, it still uses DOM for the remaining elements). Memory usage has thus drastically decreased (some caveats apply).

Closing this as completed/duplicate of #1874, because the only easy way to further decrease memory is to somehow remove the the memory used by zip file.

That is like 4 levels of abstraction deep. ClosedXML used OpenXML SDK that has OpenXmlPart that points to `OpenXmlPackage`. The `OpenXmlPackage` uses `Package` abstraction from `System.IO.Packaging` (in reality it uses an ZipPackage implementation). The `ZipPackage` uses `ZipArchive` and that finally delegates a `OpenXmlPart` stream to the `ZipArchiveEntry` stream (and there is fun with different stream wrappers). There are some knobs that can probably disable memory stream, but the knobs would need to be set correctly through all layers up to the `ZipArchiveEntry`. I looked and it haven't found a way to do it.

The hard way is to fix ClosedXML structures so it isn't such a memory hog.... not going to happen anytime soon and the problem of this ticket is additional memory required during saving anyway 😉

Duplicate of #1874
